pkg/gadgets/socket-collector/tracer/tracer.go

```diff
@@ -0,0 +1,194 @@
+		Program: objs.DumpTcp4,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach TCP BPF iterator: %v", err)
+	}
+
+	return it, nil
+}
+
+func getUDPIter() (*link.Iter, error) {
+	objs := IterUDPv4Objects{}
+	if err := LoadIterUDPv4Objects(&objs, nil); err != nil {
+		return nil, fmt.Errorf("failed to load UDP BPF objects: %v", err)
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpUdp4,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach UDP BPF iterator: %v", err)
+	}
+
+	return it, nil
+}
+
+func RunCollector(pid uint32, filter *gadgetv1alpha1.ContainerFilter) (string, error) {
+	var err error
+	var it *link.Iter
+	iters := []*link.Iter{}
+
+	// TODO: Use filters to collect only TCP or only UDP information
+	it, err = getTCPIter()
+	if err != nil {
+		return "", err
+	}
+	iters = append(iters, it)
+
+	it, err = getUDPIter()
+	if err != nil {
+		return "", err
+	}
+	iters = append(iters, it)
+
+	// Lock the OS Thread so we don't accidentally switch namespaces. Further details in:
+	// https://github.com/vishvananda/netns/blob/2eb08e3e575f00733a612d25cc5d7470f8db6f35/README.md
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// TCP/UDP BPF Iters read information from the current namespace thus
+	// we need to change network namespace before opening the iterator
+	netnsHandle, err := netns.GetFromPid(int(pid))
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve network namespace from PID %d: %v", pid, err)
+	}
+	defer netnsHandle.Close()
+
+	if err := netns.Set(netnsHandle); err != nil {
+		return "", fmt.Errorf("error changing network namespace of PID %d: %v", pid, err)
+	}
+
+	var sockets []Socket
+
+	for _, it := range iters {
+		reader, err := it.Open()
+		if err != nil {
+			return "", fmt.Errorf("failed to open BPF iterator: %v", err)
+		}
+		defer reader.Close()
+
+		scanner := bufio.NewScanner(reader)
+		for scanner.Scan() {
+			var status, proto string
+			var destp, srcp uint16
+			var dest, src uint32
+			var hexStatus uint8
+
+			len, err := fmt.Sscanf(scanner.Text(), "%s %08X %04X %08X %04X %02X", &proto, &src, &srcp, &dest, &destp, &hexStatus)
+			if err != nil || len != 6 {
+				return "", fmt.Errorf("failed to parse sockets information: %v", err)
+			}
+
+			status, err = parseStatus(proto, hexStatus)
+			if err != nil {
+				return "", err
+			}
+
+			sockets = append(sockets, Socket{
+				Protocol:            proto,
+				LocalAddress:        parseIPv4(src),
+				LocalPort:           srcp,
+				RemoteAddress:       parseIPv4(dest),
+				RemotePort:          destp,
+				Status:              status,
+				KubernetesNamespace: filter.Namespace,
+				KubernetesPod:       filter.Podname,
+			})
+		}
+
+		if err := scanner.Err(); err != nil {
+			return "", fmt.Errorf("failed reading output of BPF iterator: %v", err)
+		}
+	}
+
+	output, _ := json.MarshalIndent(sockets, "", " ")
+	return string(output), nil
```

Comment:
When there are no sockets in the pod, I get the following:
```
status:
  output: "null"
  state: Completed
```
Shouldn't we output the following instead?
```
status:
  output: "[]"
  state: Completed
```